firedrake/solving_utils.py

```diff
@@ -484,14 +485,44 @@ def compute_operators(ksp, J, P):
             return
         ctx._jacobian_assembled = True
 
-        fine = ctx._fine
-        if fine is not None:
+        # Since the finest space is not the first to be assembled, we cannot
+        # inject coefficients from the parent level without ensuring them
+        # to be in agreement with the finest level.
+
+        def get_coefficients(problem):
+            coefficients = []
+            for form in (problem.F, problem.Jp):
+                if form is not None:
+                    coefficients.extend(form.coefficients())
+            return coefficients
+
+        # TODO determine whether we are the first level to be assembled,
+        # but typically level 1 will be the first one to be assembled
+        _, l = get_level(ctx._x.function_space().mesh())
+        if l == 1:
+            # Go to the finest space
+            fine = ctx
+            while fine._fine is not None:
+                fine = fine._fine
+
+            # Inject all coefficients across the hierarchy
             manager = dmhooks.get_transfer_manager(fine._x.function_space().dm)
-            manager.inject(fine._x, ctx._x)
-
-            for bc in chain(*ctx._problem.bcs):
-                if isinstance(bc, DirichletBC):
-                    bc.apply(ctx._x)
+            while fine._coarse is not None:
+                coarse = fine._coarse
+
+                fine_coefficients = get_coefficients(fine._problem)
+                coarse_coefficients = get_coefficients(coarse._problem)
+                seen = set()
+                for xf, xc in zip(fine_coefficients, coarse_coefficients):
+                    if xf not in seen:
+                        manager.inject(xf, xc)
+                        seen.add(xf)
```

Comment:
Is it sufficient to only do this in `kspcomputeoperators` case, or should it also be done in `form_jacobian` and `form_function`? When we did this before (cc @pefarrell) we did it by setting up hooks https://github.com/firedrakeproject/firedrake/compare/master...pefarrell/transfer-manager-inject-everything

That was never merged because it does too much work (reinjects too frequently). I think this suffers from the same problem. There's no need to reinject a coefficient of `F` if it hasn't changed. These days you can ask the `dat` for its state, keep track of that and if it changed since the last time you saw it, you know its changed and you must inject the function.